File: skills/code-optimizer/references/database/python/test_code.py
from collections.abc import Callable

from code import FakeDb, Order, QueryStats, totals_bad, totals_good


def build_orders(user_count: int, orders_per_user: int) -> tuple[list[str], list[Order]]:
  user_ids: list[str] = []
  orders: list[Order] = []
  for user_index in range(user_count):
    user_id = f"user-{user_index}"
    user_ids.append(user_id)
    for order_index in range(orders_per_user):
      orders.append(
        Order(
          user_id=user_id,
          total=(order_index % 5) + 1,
        )
      )
  return user_ids, orders


def measure(
  label: str,
  work: Callable[[FakeDb, list[str]], dict[str, int]],
  db: FakeDb,
  user_ids: list[str],
) -> tuple[dict[str, int], QueryStats, str]:
  db.reset()
  result = work(db, user_ids)
  return result, db.snapshot(), label


user_ids, orders = build_orders(200, 12)
db = FakeDb(orders)
bad_result, bad_stats, bad_label = measure("totals_bad", totals_bad, db, user_ids)
good_result, good_stats, good_label = measure("totals_good", totals_good, db, user_ids)

if bad_result != good_result:
  raise RuntimeError("Database example results do not match")

for label, stats in ((bad_label, bad_stats), (good_label, good_stats)):
  print(
    label,
    {
      "query_count": stats.query_count,
      "rows_returned": stats.rows_returned,
      "bytes_returned": stats.bytes_returned,
      "elapsed_ms": stats.elapsed_ms,
    },
  )

File: skills/code-optimizer/references/database/node/test_code.ts
import { FakeDb, totalsBad, totalsGood, type Order, type QueryStats } from "./code.ts";

type Sample = QueryStats & {
  label: string;
  result: Array<[string, number]>;
};

function buildOrders(
  userCount: number,
  ordersPerUser: number,
): { userIds: string[]; orders: Order[] } {
  const userIds: string[] = [];
  const orders: Order[] = [];
  for (let userIndex = 0; userIndex < userCount; userIndex += 1) {
    const userId = `user-${userIndex}`;
    userIds.push(userId);
    for (let orderIndex = 0; orderIndex < ordersPerUser; orderIndex += 1) {
      orders.push({
        userId,
        total: (orderIndex % 5) + 1,
      });
    }
  }
  return { userIds, orders };
}

function measure(
  label: string,
  work: (db: FakeDb, userIds: string[]) => Map<string, number>,
  db: FakeDb,
  userIds: string[],
): Sample {
  db.reset();
  const result = work(db, userIds);
  return {
    label,
    result: Array.from(result.entries()),
    ...db.snapshot(),
  };
}

const { userIds, orders } = buildOrders(200, 12);
const db = new FakeDb(orders);
const bad = measure("totalsBad", totalsBad, db, userIds);
const good = measure("totalsGood", totalsGood, db, userIds);

if (JSON.stringify(bad.result) !== JSON.stringify(good.result)) {
  throw new Error("Database example results do not match");
}

console.table([
  {
    label: bad.label,
    queryCount: bad.queryCount,
    rowsReturned: bad.rowsReturned,
    bytesReturned: bad.bytesReturned,
    elapsedMs: bad.elapsedMs,
  },
  {
    label: good.label,
    queryCount: good.queryCount,
    rowsReturned: good.rowsReturned,
    bytesReturned: good.bytesReturned,
    elapsedMs: good.elapsedMs,
  },
]);
